File: flyweight/java/FoodFactory.java
package flyweight.java;

import java.util.Map;

public class FoodFactory {
    static Map<String, FoodType> foodTypes = new java.util.HashMap<String, FoodType>();

    public static FoodType getFoodType(String cuisine, String category) {
        String key = cuisine + "-" + category;
        if (!foodTypes.containsKey(key)) {
            foodTypes.put(key, new FoodType(cuisine, category));
        }
        return foodTypes.get(key);
    }
}

File: flyweight/golang/foodfactory.go
package main

type FoodFactory struct {
	foodTypes map[string]FoodType
}

func NewFoodFactory() *FoodFactory {
	return &FoodFactory{
		foodTypes: make(map[string]FoodType),
	}
}

func (ff *FoodFactory) GetFoodType(cusine string, category string) FoodType {
	foodTypeName := cusine + "-" + category
	if foodType, exists := ff.foodTypes[foodTypeName]; exists {
		return foodType
	}
	var foodType FoodType = NewFoodType(cusine, category)
	ff.foodTypes[foodTypeName] = foodType
	return foodType
}
